### PR TITLE
Spool items that weren't stored before

### PR DESCRIPTION
**Pull request by annando**
Some times a day I see entries in my log file that an item couldn't be stored due to a deadlock. The new coding now stores these items into a spool directory. The files in that directory will then be posted some time later.

### CODE REVIEW

File: boot.php [hunk 1]
```diff
@@ -2365,6 +2365,29 @@ function get_lockpath() {
 	return "";
 }
 
+function get_spoolpath() {
```
Comment (rabuzarus) [inline boot.php]: please add doxygen description

File: boot.php [hunk 2]
```diff
@@ -2365,6 +2365,29 @@ function get_lockpath() {
...
+	$spoolpath = get_config('system','spoolpath');
+	if (($spoolpath != "") AND is_dir($spoolpath) AND is_writable($spoolpath))
```
Comment (rabuzarus) [inline boot.php]: STANDARDS - please use brackets

File: boot.php [hunk 3]
```diff
@@ -2365,6 +2365,29 @@ function get_lockpath() {
...
+		return($spoolpath);
+
+	$temppath = get_temppath();
+
+	if ($temppath != "") {
+		$spoolpath = $temppath."/spool";
+
+		if (!is_dir($spoolpath))
```
Comment (rabuzarus) [inline boot.php]: STANDARDS - please use brackets

File: include/items.php [hunk 4]
```diff
@@ -845,15 +846,26 @@ function item_store($arr,$force_parent = false, $notify = false, $dontcache = fa
...
 	} else {
 		// This can happen - for example - if there are locking timeouts.
-		logger("Item wasn't stored - we quit here.");
-		q("COMMIT");
+		q("ROLLBACK");
+
+		// Store the data into a spool file so that we can try again later.
+
+		// At first we restore the Diaspora signature that we removed above.
+		if (isset($encoded_signature))
```
Comment (rabuzarus) [inline include/items.php]: STANDARDS - please use brackets

File: include/items.php [hunk 5]
```diff
@@ -845,15 +846,26 @@ function item_store($arr,$force_parent = false, $notify = false, $dontcache = fa
 		}
 	} else {
 		// This can happen - for example - if there are locking timeouts.
-		logger("Item wasn't stored - we quit here.");
-		q("COMMIT");
+		q("ROLLBACK");
+
+		// Store the data into a spool file so that we can try again later.
```
Comment (rabuzarus) [inline include/items.php]: not a showstopper but I want to give the suggestion that we should start to create separate functions for some of its parts. `function item_store` is becoming bigger and bigger. Maybe we should think about a own storing class

File: boot.php (hunk 2 above)
Comment (annando) [inline boot.php]: Am I allowed to call you a second @hypolite? :-)

File: include/items.php (hunk 5 above)
Comment (Hypolite) [inline include/items.php]: I'd definitely could assist you in designing a class architecture. I would have to spend a good deal of time studying the current item functions however. So far I only scanned them and changed the parts that were only directly relevant to the issue I was interested in. I'll try to come up with a little design document to make sure I'm not forgetting anything.

File: boot.php (hunk 2 above)
Comment (Hypolite) [inline boot.php]: Only if you mean it as flattering! 😄

File: include/items.php (hunk 5 above)
Comment (annando) [inline include/items.php]: I'm not sure if you really need a deep look into the current functions. I would need some skeleton, then I could do the work. I can write down which functions are needed in general.

Important is the fetching. I would like to replace the various database queries at many places with function calls.

**Pull request merged**